ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java

```diff
@@ -590,140 +590,270 @@ private boolean copyDataByIndexSegment(LogSegment logSegmentToCopy, IndexSegment
-            validEntries.add(indexEntry);
+      } else {
+        IndexValue valueFromIdx = srcIndex.findKey(indexEntry.getKey());
+        // Doesn't matter whether we get the PUT or DELETE entry for the expiry test
+        if (!srcIndex.isExpired(valueFromIdx)) {
+          // unexpired PUT entry.
+          if (deletesInEffect) {
+            if (!isDeleted(indexEntry.getKey(), indexSegment.getStartOffset(),
+                startOffsetOfLastIndexSegmentForDeleteCheck)) {
+              // PUT entry that has not expired and is not considered deleted.
+              // Add all values in this index segment (to account for the presence of TTL updates)
+              addAllEntriesForKeyInSegment(validEntries, indexSegment, indexEntry);
+            } else {
+              logger.trace("{} in index segment with start offset {} in {} is not valid because it is a deleted PUT",
+                  indexEntry, indexSegment.getStartOffset(), storeId);
+            }
           } else {
-            logger.trace("Skipping {} in index segment with start offset {} in {} because it is a deleted PUT",
-                indexEntry, indexSegmentStartOffset, storeId);
+            // valid PUT entry
+            // Add all values in this index segment (to account for the presence of TTL updates)
+            addAllEntriesForKeyInSegment(validEntries, indexSegment, indexEntry);
           }
         } else {
-          // valid PUT entry
-          validEntries.add(indexEntry);
+          logger.trace("{} in index segment with start offset {} in {} is not valid because it is an expired PUT",
+              indexEntry, indexSegment.getStartOffset(), storeId);
         }
-      } else {
-        logger.trace("Skipping {} in index segment with start offset {} in {} because it is an expired PUT", indexEntry,
-            indexSegmentStartOffset, storeId);
       }
     }
     return validEntries;
   }
 
   /**
-   * Checks if a record already exists in the target log (swap spaces).
+   * Determines whether a TTL update entry is valid
+   * @param key the {@link StoreKey} being examined
+   * @param indexSegmentStartOffset the start offset of the {@link IndexSegment} that the TTL update record is in
+   * @return {@code true} if the TTL update entry is valid
+   * @throws StoreException if there are problems reading the index
+   */
+  private boolean isTtlUpdateEntryValid(StoreKey key, Offset indexSegmentStartOffset) throws StoreException {
+    boolean valid = false;
+    //  A TTL update entry is "valid" if the corresponding PUT is still alive
+    // The PUT entry, if it exists, must be "before" this TTL update entry.
+    FileSpan srcSearchSpan = new FileSpan(srcIndex.getStartOffset(), indexSegmentStartOffset);
+    IndexValue srcValue = srcIndex.findKey(key, srcSearchSpan, EnumSet.of(PersistentIndex.IndexEntryType.PUT));
+    if (srcValue == null) {
+      // PUT is not in the source - therefore can't be in target. This TTL update can be cleaned up
```

Comment:
when you say "therefore cannot be in target", do you mean "therefore cannot go into the target"? (just a little confusing).